### PR TITLE
Basic tls support 2

### PR DESCRIPTION
**Pull request by mjgarton**
Second attempt at TLS support. See individual commit message for details.

### CODE REVIEW

**Comment by jonhoo** [comment]
Thanks! I'm happy with this approach, so we can close the other PR, though see https://github.com/jonhoo/msql-srv/pull/26#issuecomment-903188116.

File: Cargo.toml
Comment (jonhoo) [inline Cargo.toml:31]: Hmm, why do we need the beta here?

File: src/commands.rs
Comment (jonhoo) [inline src/commands.rs:29]: Is it not possible to get a username with `CLIENT_SSL`?

File: src/lib.rs
Comment (jonhoo) [inline src/lib.rs:233]: The documentation for this method will need to be updated.

File: src/lib.rs
Comment (jonhoo) [inline src/lib.rs:194]: Is there a sensible default configuration we can supply here?

File: src/packet.rs
Comment (jonhoo) [inline src/packet.rs:72]: Can this actually happen "by accident" in production? If so, we should handle this case more gracefully.

File: src/tls.rs
Comment (jonhoo) [inline src/tls.rs:17]: It seems unfortunate for us to hard-code these things here. Can we instead let the user provide us with a `rustls::ServerConfig` directly?

File: src/tls.rs
Comment (jonhoo) [inline src/tls.rs:43]: This feels like something that shouldn't live in this library.

File: src/tls.rs
Comment (jonhoo) [inline src/tls.rs:79]: I don't follow why we need this wrapper?

File: src/packet.rs
Comment (mjgarton) [reply src/packet.rs:72]: If this happens, it would represent a bug. I don't expect it ever could happen, and perhaps can be removed. I will remove it if you prefer.

File: src/tls.rs
Comment (mjgarton) [reply src/tls.rs:17]: Yes. I think I was trying to avoid exposing the fact that we're using rustls, but this is worse.  I'll change this.

File: src/tls.rs
Comment (mjgarton) [reply src/tls.rs:43]: I agree.  Similar to above, I was trying to avoid exposing the fact that we're using rustls.  But this code clearly smells of my particular use cases, so I'll update this.

File: src/packet.rs
Comment (jonhoo) [reply src/packet.rs:72]: Fair enough — let's leave it as an assertion for now then!

File: src/lib.rs
Comment (mjgarton) [reply src/lib.rs:193]: I am not sure there is. Also, `None` really means "don't support TLS" which is a reasonable default in itself I think.

Comment (jonhoo) [reply src/lib.rs:193]: I guess that's the part I'm pushing back on — as a user, I would expect TLS to work "out of the box", as it does with basically all clients these days, using the system credentials. Presumably `rustls` has a default TLS config too?

Comment (jonhoo) [reply src/lib.rs:193]: Oh, wait, I'm being dumb. This is a server, not a client, so configuration is definitely non-trivial, needing a server certificate and such. Ignore me :sweat_smile:

File: Cargo.toml
Comment (mjgarton) [reply Cargo.toml:31]: I need `0.20.0` because I can't see how to achieve what I want in earlier versions.  Even if it's somehow possible, I'd like to avoid the rework.    If you are happy with it, the best approach might be to hold off merging this until `0.20.0` is released, which doesn't seem far off. See https://github.com/rustls/rustls/issues/813

**Comment by sundy-li** [comment]
LGTM, I want to implement AuthPlugin. Seems `caching_sha2_password` depends on SSL connection.

File: Cargo.toml
Comment (mjgarton) [reply Cargo.toml]: I bumped to 0.20.0 final, so I think this PR is ready now @jonhoo

File: Cargo.toml
Comment (jonhoo) [inline Cargo.toml:32]: I don't think this is needed any more?

File: src/lib.rs
Comment (jonhoo) [inline src/lib.rs:192]: Since what we ultimately need is `Arc<rustls::ServerConfig>`, let's have the caller directly provide that. That way, they can re-use a single `Arc` rather than us creating a new one for each connection. It _also_ allows them to create configs on-the-fly, which the current configuration does not.

**Review by jonhoo — CHANGES_REQUESTED** [review]
Just two minor comments this time around!

One thing I would still like to see before landing this though is one or two tests that all this new logic actually works correctly. The easiest way is probably to require checking in a self-signed cert (or maybe generating one on the fly for the test), and then configuring the MySQL client in the test to not require a valid cert (or seed it with the public key for a more reliable test).

File: Cargo.toml
Comment (mjgarton) [reply Cargo.toml:32]: I will remove this.

File: src/lib.rs
Comment (mjgarton) [reply src/lib.rs]: Fixed.

**Comment by mjgarton** [comment]
@jonhoo I finally found time to get back to finishing this off.  It should be ready now (including tests)

(Assuming you are happy and merge this, I would like to let you know that I am likely to create another separate PR with additional new functionality and with further breaking changes, probably within the next week.  So if you wish to bear that in mind for the timing of any new release, please do so.)

File: src/lib.rs
Comment (jonhoo) [inline src/lib.rs:249]: It seems possibly unfortunate that we're calling this multiple times, sometimes throwing away the result. I wonder if we should just call `.tls_config()` in `::new` and then save it so we can then use it in places like here. What do you think?

File: src/packet.rs
Comment (jonhoo) [inline src/packet.rs:72]: I missed this last time around — this should use `assert_eq!` so that if it fails we learn how much _was_ remaining,

File: src/packet.rs
Comment (jonhoo) [inline src/packet.rs:217]: I don't think this needs to be `pub`? Or am I missing something? It can probably just be `pub(crate)`.

File: src/packet.rs
Comment (jonhoo) [inline src/packet.rs:219]: Same here — probably doesn't need to be `pub`.

File: tests/main.rs
Comment (jonhoo) [inline tests/main.rs:222]: nit: you can use `Result::expect_err` for this to make the code a little nicer.

File: Cargo.toml
Comment (jonhoo) [inline Cargo.toml:31]: I wonder if we should make TLS support optional to save people some compile time. Basically an on-by-default feature that we gate all the TLS stuff behind.

**Review by jonhoo — CHANGES_REQUESTED** [review]
Thanks for finding the time to look at this again! Went through and found some more bits, but nothing too major.

File: src/packet.rs
Comment (mjgarton) [reply src/packet.rs]: Done.

File: src/lib.rs
Comment (mjgarton) [reply src/lib.rs]: I agree. Done & pushed.

File: Cargo.toml
Comment (mjgarton) [reply Cargo.toml]: Done.

**Comment by mjgarton** [comment]
@jonhoo Please don't merge this yet. I have found a minor bug (that only happens when TLS is enabled) that I am preparing a fix and a test for.

File: tests/main.rs
Comment (jonhoo) [inline tests/main.rs:186]: This seems like a spurious newline.

File: tests/main.rs
Comment (jonhoo) [inline tests/main.rs:201]: This seems like a spurious newline.

**Review by jonhoo — APPROVED** [review]
I'm happy with this once we remove the two stray newlines! :tada:

**Comment by jonhoo** [comment]
I'd still love for us to find a way to avoid copying the packet bytes each time, but feels like a longer-term TODO than something that needs to be fixed in this PR.

**Comment by jonhoo** [comment]
It looks like there's a conflict with the main branch?

**Comment by jonhoo** [comment]
Merged! :tada: 
Thanks for sticking with this for so long :)

**Pull request merged**
